File: counter/app/src/main/java/com/example/counter/MainActivity.java
package com.example.counter;

import androidx.appcompat.app.AppCompatActivity;
import android.os.Bundle;
import android.os.Handler;
import android.view.View;
import android.widget.Button;
import android.widget.TextView;
public class MainActivity extends AppCompatActivity {
    Button btnstart,btnstop;
    TextView textcount;
    int i=1;
    Handler newhandler = new Handler();
    @Override
    protected void onCreate(Bundle savedInstanceState) {  super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        btnstart = findViewById(R.id.buttonstart); btnstop = findViewById(R.id.buttonstop);
        textcount = findViewById(R.id.txtcount);
        btnstart.setOnClickListener(new View.OnClickListener() {  @Override
        public void onClick(View v) {
            newhandler.postDelayed(upTimerthread,0);  }
        });
        btnstop.setOnClickListener(new View.OnClickListener() {  @Override
        public void onClick(View v) {
            newhandler.removeCallbacks(upTimerthread);  }
        });
    }
    private Runnable upTimerthread = new Runnable() {  @Override
    public void run() {
        textcount.setText(""+i);  newhandler.postDelayed(this,1000);  i++;
    }
    };
}

File: TipTime/app/src/main/java/com/example/tiptime/MainActivity.kt
package com.example.tiptime

import android.content.Context
import androidx.appcompat.app.AppCompatActivity
import android.os.Bundle
import android.view.KeyEvent
import android.view.View
import android.view.inputmethod.InputMethodManager
import android.widget.TextView
import com.example.tiptime.databinding.ActivityMainBinding
import java.text.NumberFormat
import kotlin.math.ceil
class MainActivity : AppCompatActivity() {

    private lateinit var binding: ActivityMainBinding

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = ActivityMainBinding.inflate(layoutInflater)
        setContentView(this.binding.root)

        binding.calculateButton.setOnClickListener {
            calculateTip()
        }
        binding.costOfServiceEditText.setOnKeyListener { view, keyCode, _ -> handleKeyEvent(view, keyCode)
        }
    }

    private fun calculateTip() {
        val stringInTextField = binding.costOfServiceEditText.text.toString()
        val cost = stringInTextField.toDoubleOrNull()
        if(cost==null||cost==0.0) {
            displayTIp(0.0)
            return
        }
        val tipPercent=when(binding.tipOptions.checkedRadioButtonId)
        {
            R.id.amazing->0.20
            R.id.good->0.18
            else->0.15
        }
        var tip=cost*tipPercent
        if(binding.switch1.isChecked)
            tip= ceil(tip)

       displayTIp(tip)
    }
    private fun displayTIp(tip:Double)
    {
        val formattedTip=NumberFormat.getCurrencyInstance().format(tip)
        binding.tipResult.text=getString(R.string.tip_amount,formattedTip)
    }
    private fun handleKeyEvent(view: View, keyCode: Int): Boolean {
        if (keyCode == KeyEvent.KEYCODE_ENTER) {
            // Hide the keyboard
            val inputMethodManager =
                getSystemService(Context.INPUT_METHOD_SERVICE) as InputMethodManager
            inputMethodManager.hideSoftInputFromWindow(view.windowToken, 0)
            return true
        }
        return false
    }

}